subsys/net/ip/net_pkt.c

```diff
@@ -1177,23 +1211,37 @@ int net_pkt_alloc_buffer(struct net_pkt *pkt,
 	}
 
 #if NET_LOG_LEVEL >= LOG_LEVEL_DBG
-	buf = pkt_alloc_buffer(pool, alloc_len, timeout, caller, line);
+	buf = pkt_alloc_buffer(pool, alloc_len, reserve_ll_header,
+			       timeout, caller, line);
 #else
-	buf = pkt_alloc_buffer(pool, alloc_len, timeout);
+	buf = pkt_alloc_buffer(pool, alloc_len, reserve_ll_header, timeout);
 #endif
 
 	if (!buf) {
 #if NET_LOG_LEVEL >= LOG_LEVEL_DBG
 		NET_ERR("Data buffer (%zd) allocation failed (%s:%d)",
-			alloc_len, caller, line);
+			alloc_len + reserve_ll_header, caller, line);
 #else
-		NET_ERR("Data buffer (%zd) allocation failed.", alloc_len);
+		NET_ERR("Data buffer (%zd) allocation failed.",
+			alloc_len + reserve_ll_header);
 #endif
 		return -ENOMEM;
 	}
 
 	net_pkt_append_buffer(pkt, buf);
 
+	/* Hide the link layer header for now. The space is used when
+	 * link layer header needs to be written to the packet by L2 send.
+	 */
+	if (reserve_ll_header > 0U) {
+		NET_DBG("Reserving %zd bytes for L2 header", reserve_ll_header);
+
+		net_buf_add(pkt->buffer, reserve_ll_header);
+		net_buf_pull(pkt->buffer, reserve_ll_header);
```

Comment:
Wouldn't it be more fit to use dedicated `net_buf_reserve()` here? The final outcome should be the same, data pointer moved by the reserved space length.

Reply:
Sure, we can do it like that too.